### PR TITLE
REGR: Fix fillna making a copy when dict was given as fill value and inplace is set

### LINKED ISSUE #47188: BUG: inplace behavior is inconsist for fillna
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
pdf = pd.DataFrame({"x": [np.nan, 2, 3, 4, np.nan, 6], "y": [np.nan, 2, 3, 4, np.nan, 6]})
pser = pdf.x
# Didn't make a copy write to existing  (it's unexpected behavior)
pdf.fillna(0, inplace=True)
>>> pser
0    0.0
1    2.0
2    3.0
3    4.0
4    0.0
5    6.0


import pandas as pd
import numpy as np
pdf = pd.DataFrame({"x": [np.nan, 2, 3, 4, np.nan, 6], "y": [np.nan, 2, 3, 4, np.nan, 6]})
pser = pdf.x
# Make a copy and doesn't write to existing (it's expected behavior)
>>> pdf.fillna({"x": -1, "y": -2}, inplace=True)
>>> pser
0    NaN
1    2.0
2    3.0
3    4.0
4    NaN
5    6.0
Name: x, dtype: float64
```

### Issue Description

Looks like behavior is a little bit differet in current pandas. 

### Expected Behavior

~Always make a copy and doesn't write to existing array for~ Do in-place operation in all functions **with inplace** (such like  `eval`/`update`/`fillna`).

### Installed Versions

<details>

1.4+

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #47188 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I think updating inplace is sensible here? This is a subtle change so I think we should backport, because the fact that we are using setitem under th hood should not leak into the interface. If we want to change this, we should deprecate, but only makes sens if we don't get Copy on Write in for 2.0

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
> I think updating inplace is sensible here? This is a subtle change so I think we should backport, because the fact that we are using setitem under th hood should not leak into the interface. If we want to change this, we should deprecate, but only makes sens if we don't get Copy on Write in for 2.0

@phofl can you also add your comments to the issue, #47188 to avoid having the discussion in more than one place as I think the expected behavior has not yet been confirmed.

**Comment by phofl** [comment]
Done

**Comment by simonjayhawkins** [comment]
> Done

Thanks. The fix here is different to suggested by @jbrockmendel in https://github.com/pandas-dev/pandas/issues/47188#issuecomment-1143702013 so requesting a review from Brock.

**Comment by phofl** [comment]
Yep, sure.

Somehow this fails on the ci while passing locally, will have to investigate

**Comment by jbrockmendel** [comment]
Makes sense to me.  The solution I suggested in #47188 might avoid creating an intermediate array, but I think without the solution here it would be a wash.  So it might be worth looking into combining the two at some point.

**Review by jbrockmendel — APPROVED** [review]
LGTM pending green

**Comment by jreback** [comment]
hmm these are failing actually: https://github.com/pandas-dev/pandas/runs/6858133053?check_suite_focus=true

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1183,6 +1183,9 @@ def fillna(
             #  test_fillna_dtype_conversion_equiv_replace
             nbs = self.where(value, ~mask.T, _downcast=False)
 
+        if inplace:
+            return nbs
```
Comment (simonjayhawkins) [inline pandas/core/internals/blocks.py:1187]: can just return early in the `if inplace:` block above?

**Comment by phofl** [comment]
Failure is the same, was just interested if it would make any difference.

Tried setting up a new env and tried editing 1.4.2 release, but can not reproduce the failure...

**Comment by simonjayhawkins** [comment]
> but can not reproduce the failure...

some of the failures are for duplicate labels. maybe loc and setitem handle them differently, although that doesn't explain why it works on one platform and not another.

**Comment by phofl** [comment]
I think the duplicated failures are just deprecation warnings, which is expected I think.

What really bugs me is the new test that fails

**Comment by simonjayhawkins** [comment]
> What really bugs me is the new test that fails

yeah, difficult without being able to reproduce. maybe just use the ci for troubleshooting and mark this as draft while debugging.

**Comment by phofl** [comment]
This is really tricky, don't really have an idea where to start, because this should work

**Comment by simonjayhawkins** [comment]
> This is really tricky, don't really have an idea where to start, because this should work

it's failing on the array manager tests? can reproduce with

```
PANDAS_DATA_MANAGER=array pytest pandas/tests/frame/methods/test_fillna.py -k test_inplace_dict_update_view
```

can probably just skip them.

**Comment by phofl** [comment]
There is something really weird going on. I skipped the test for the array manager which removed one of the failures. This looks to me, as somehow array manager is set

**Comment by phofl** [comment]
Yep definitely should remove. Not quite finished yet. As @yikun mentioned, should we also change for eval and update? Could do this in one pr.

**Comment by simonjayhawkins** [comment]
> Could do this in one pr.

sure.

**Comment by phofl** [comment]
Removed eval, this is tricky. Update was straightforward

**Review by simonjayhawkins — APPROVED** [review]

**Comment by jreback** [comment]
doesn't this overlap with the PR for the performance change in update?

**Comment by phofl** [comment]
Good point, forgot about that.

This restores behavior from 1.3.5, so the other pr was not only about performance, it would have also tackled a regression. Did not think about that before

**Comment by simonjayhawkins** [comment]
merge this and close #47407?

**Comment by phofl** [comment]
Yes, if @jreback is ok?

I would open a dedicated issue for eval, the other is already pretty long

**Comment by simonjayhawkins** [comment]
> I would open a dedicated issue for eval, the other is already pretty long

#47449

**Comment by phofl** [comment]
Thx

**Pull request merged**